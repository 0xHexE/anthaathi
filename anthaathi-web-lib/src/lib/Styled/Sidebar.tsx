import * as React from 'react';
import { styled, Theme, useStyletron } from 'baseui';
import { Button, SHAPE } from 'baseui/button';
import { PLACEMENT, StatefulTooltip } from 'baseui/tooltip';
import { useAtom } from 'jotai';
// @ts-ignore TODO: FIX IT
import { ChevronRight } from '@carbon/icons-react';
import Anthaathi from '../Icon/Anthaathi';
import SidebarOpen from '../Atoms/SidebarOpen';

const Sidebar = styled(
  'div',
  ({ $theme, $expanded }: { $theme?: Theme; $expanded?: boolean }) => ({
    display: 'flex',
    minHeight: `calc(100vh - ${$theme!.sizing.scale400} - ${
      $theme!.sizing.scale400
    })`,
    flexDirection: 'column',
    width: $expanded ? '320px' : '64px',
    backgroundColor: $theme?.colors.primary700,
    alignItems: 'center',
    transition: `width ${$theme!.animation.easeInCurve} ${
      $theme!.animation.timing200
    }`,
    paddingTop: $theme?.sizing.scale400,
    paddingBottom: $theme?.sizing.scale400,
  }),
);

export const SidebarDivider = styled('div', ({ $theme }) => ({
  width: `calc(100% - ${$theme.sizing.scale600})`,
  backgroundColor: $theme.colors.backgroundTertiary,
  marginTop: $theme.sizing.scale100,
  marginBottom: $theme.sizing.scale100,
  opacity: '.2',
  height: '1px',
}));

export interface SidebarIconButtonProps {
  isOpen: boolean;
  icon?: React.ReactElement;
  title: React.ReactNode;
  tooltip: React.ReactNode;
  iconClass?: string;
  onClick?: () => void;
}

export function SidebarIconButton({
  isOpen,
  icon: icon_,
  title,
  tooltip,
  iconClass,
  onClick,
}: SidebarIconButtonProps) {
  const [css, $theme] = useStyletron();

  const icon = React.cloneElement(icon_ || <Anthaathi />, {
    width: 48,
    className: `${css({
      overflow: 'inherit',
      width: '48px',
      height: '48px',
    })} ${iconClass}`.trim(),
  });

  return (
    <StatefulTooltip
      content={tooltip}
      popoverMargin={18}
      placement={$theme.direction === 'rtl' ? PLACEMENT.left : PLACEMENT.right}
    >
      <Button
        shape={SHAPE.square}
        $style={{
          backgroundColor: 'transparent',
          width: `calc(100% - ${$theme.sizing.scale600})`,
          display: 'flex',
          cursor: onClick ? 'pointer' : 'default',
          placeContent: 'start',
          alignItems: 'center',
          ':active': {
            ...(onClick
              ? {}
              : {
                  backgroundColor: 'transparent',
                }),
          },
        }}
        onClick={onClick}
      >
        {icon}

        <span
          className={css({
            overflow: 'hidden',
            opacity: isOpen ? 1 : 0,
            transition: `transform ${$theme!.animation.easeInCurve} ${
              $theme!.animation.timing200
            }, opacity ${$theme!.animation.easeInCurve} ${
              $theme!.animation.timing200
            }`,
            display: 'block',
            whiteSpace: 'nowrap',
          })}
        >
          {title}
        </span>
      </Button>
    </StatefulTooltip>
  );
}

export function SidebarToggle() {
  const [isOpen, setIsOpen] = useAtom(SidebarOpen);
  const [css, $theme] = useStyletron();

  return (
    <Button
      $style={{
        width: `calc(100% - ${$theme.sizing.scale600})`,
        backgroundColor: 'transparent',
      }}
      onClick={() => setIsOpen((prev) => !prev)}
    >
      <ChevronRight
        className={css({
          transform: isOpen ? 'rotate(180deg)' : 'rotate(0deg)',
          transition: `transform ${$theme.animation.easeInCurve} ${$theme.animation.timing200}`,
        })}
        size={18}
      />
    </Button>
  );
}

export default Sidebar;
