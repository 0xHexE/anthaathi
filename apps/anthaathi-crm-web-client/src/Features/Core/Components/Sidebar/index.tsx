import { styled } from 'baseui';

export const Sidebar = styled('div', ({ $theme }) => ({
  width: '320px',
  position: 'fixed',
  left: 0,
  top: '48px',
  height: 'calc(100vh - 48px)',
  backgroundColor: '#2e3439',
  boxShadow: $theme.lighting.shadow400,
  transitionProperty: 'all',
  transitionDuration: '100ms',
  transitionTimingFunction: 'ease',
  zIndex: 1,
}));
