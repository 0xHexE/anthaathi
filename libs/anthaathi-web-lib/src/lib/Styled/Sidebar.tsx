import * as React from 'react';
import { useCallback, useEffect, useRef, useState } from 'react';
import { styled, useStyletron } from 'baseui';
import { LabelSmall } from 'baseui/typography';
// @ts-ignore
import { ChevronUp } from '@carbon/icons-react';
import { FlexFill } from './Utils';

const animationTiming = process.env.NODE_ENV === 'test' ? 0 : 220;

const Sidebar = styled('div', {
  width: '320px',
  backgroundColor: '#EEE',
  height: '100vh',
  overflow: 'auto',
});

export const SidebarToolbar = styled('h1', ({ $theme }) => ({
  backgroundColor: $theme.colors.backgroundAccent,
  height: '64px',
  display: 'flex',
  alignItems: 'center',
  paddingLeft: $theme.sizing.scale800,
  paddingRight: $theme.sizing.scale800,
  color: $theme.colors.contentInversePrimary,
  marginTop: 0,
  marginBottom: 0,
  width: `calc(100% - ${$theme.sizing.scale800} - ${$theme.sizing.scale800})`,
  ...$theme.typography.HeadingXSmall,
}));

export interface SidebarItemProps {
  title: React.ReactNode;
  children: React.ReactNode;
  action?: React.ReactNode;
}

export function SidebarItem({ title, children, action }: SidebarItemProps) {
  const [css, $theme] = useStyletron();

  const [isOpen, setIsOpen] = useState(false);

  const actionRef = useRef<HTMLSpanElement>(null);

  const handleClick = useCallback(
    (
      e: React.KeyboardEvent<HTMLDivElement> | React.MouseEvent<HTMLDivElement>,
    ) => {
      if ((e.target as HTMLDivElement).closest('[data-kind="action"]')) {
        return;
      }

      setIsOpen((prev) => !prev);
    },
    [],
  );

  const [shouldRender, setShouldRender] = useState(false);

  // eslint-disable-next-line no-undef
  const ref = useRef<NodeJS.Timeout>();

  useEffect(() => {
    if (isOpen) {
      setShouldRender(true);
    } else {
      if (ref.current) {
        clearTimeout(ref.current);
      }

      ref.current = setTimeout(() => {
        setShouldRender(false);
      }, animationTiming);
    }
  }, [isOpen]);

  return (
    <>
      <div
        tabIndex={0}
        role="button"
        data-testid="sidebar-title"
        aria-expanded={isOpen ? 'true' : 'false'}
        onKeyDown={(e) => {
          if (!(e.key === 'Enter' || e.code === 'Enter')) {
            return;
          }
          handleClick(e);
        }}
        onClick={(e) => {
          handleClick(e);
        }}
        className={css({
          display: 'flex',
          alignItems: 'center',
          height: '32px',
          paddingLeft: $theme.sizing.scale600,
          paddingRight: $theme.sizing.scale600,
          cursor: 'pointer',
          userSelect: 'none',
          transition: `background-color ${$theme.animation.easeInCurve} ${$theme.animation.timing100}`,
          ':hover': {
            backgroundColor: $theme.colors.backgroundSecondary,
          },
        })}
      >
        <LabelSmall>{title}</LabelSmall>

        <span className={css({ flexGrow: 1 })} />

        <span ref={actionRef} data-kind="action">
          {action}
        </span>

        <ChevronUp
          className={css({
            transform: isOpen ? 'rotate(0)' : 'rotate(180deg)',
            marginLeft: '12px',
          })}
        />
      </div>
      <div
        className={css({
          paddingTop: isOpen ? $theme.sizing.scale400 : 0,
          paddingBottom: isOpen ? $theme.sizing.scale400 : 0,
          paddingLeft: 0,
          paddingRight: 0,
          background: $theme.colors.backgroundPrimary,
          overflow: 'hidden',
          maxHeight: isOpen ? '1000px' : '0',
          transition: `padding ease ${$theme.animation.timing200}, max-height ease ${$theme.animation.timing200}`,
        })}
        tabIndex={isOpen ? -1 : undefined}
      >
        {shouldRender && children}
      </div>
    </>
  );
}

export interface SidebarItemLinkProps
  extends React.HTMLAttributes<HTMLDivElement> {
  children: React.ReactNode;
  icon?: React.ReactNode;
  action?: React.ReactNode;
  isHeading?: boolean;
}

export function SidebarItemLink({
  children,
  icon,
  onClick,
  isHeading = false,
  action,
  ...props
}: SidebarItemLinkProps) {
  const [css, $theme] = useStyletron();

  return (
    <div
      role="link"
      tabIndex={0}
      onKeyDown={(e) => {
        if (!(e.key === 'Enter' || e.code === 'Enter')) {
          return;
        }

        onClick?.(e as never);
      }}
      onClick={onClick as never}
      className={css({
        cursor: 'pointer',
        ':hover': {
          backgroundColor: $theme.colors.backgroundSecondary,
        },
        height: isHeading ? '32px' : '48px',
        display: 'flex',
        alignItems: 'center',
        userSelect: 'none',
        paddingLeft: $theme.sizing.scale600,
        paddingRight: $theme.sizing.scale600,
        color: isHeading
          ? $theme.colors.backgroundSecondary
          : $theme.colors.primaryA,
      })}
      {...props}
    >
      {icon && (
        <div
          className={css({
            width: '18px',
            height: '18px',
            marginRight: $theme.sizing.scale300,
          })}
        >
          {icon}
        </div>
      )}
      <LabelSmall>{children}</LabelSmall>
      {action && <FlexFill />}
      {action}
    </div>
  );
}

export default Sidebar;
