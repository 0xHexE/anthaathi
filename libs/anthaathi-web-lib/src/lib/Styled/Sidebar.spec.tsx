import * as React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { SidebarItem, SidebarItemLink } from './Sidebar';

describe('Sidebar', () => {
  it('should render the sidebar', async () => {
    const items = render(
      <SidebarItem
        title="title"
        action={
          <button type="button" data-testid="hello">
            Hello world
          </button>
        }
      >
        test
      </SidebarItem>,
    );

    userEvent.click(items.getByTestId('hello'));

    expect(items.getByTestId('sidebar-title')).toMatchSnapshot();

    expect(
      items
        .getByTestId('sidebar-title')
        .attributes.getNamedItem('aria-expanded')?.value,
    ).toBe('false');

    userEvent.click(items.getByTestId('sidebar-title'));

    expect(
      items
        .getByTestId('sidebar-title')
        .attributes.getNamedItem('aria-expanded')?.value,
    ).toBe('true');
  });

  it('should expand when press enter', () => {
    const func = jest.fn().mockImplementation(() => {});

    const items = render(
      <SidebarItem
        title="title"
        action={
          <button type="button" data-testid="hello">
            Hello world
          </button>
        }
      >
        <SidebarItemLink data-testid="tobeclicked" onClick={func}>
          TTest
        </SidebarItemLink>
      </SidebarItem>,
    );

    const ele = items.getByTestId('sidebar-title');

    ele.focus();

    userEvent.keyboard('e');

    expect(
      items
        .getByTestId('sidebar-title')
        .attributes.getNamedItem('aria-expanded')?.value,
    ).toBe('false');

    userEvent.keyboard('{enter}');

    expect(
      items
        .getByTestId('sidebar-title')
        .attributes.getNamedItem('aria-expanded')?.value,
    ).toBe('true');

    items.getByTestId('tobeclicked').focus();

    userEvent.keyboard('{enter}');

    expect(func).toBeCalledTimes(1);

    userEvent.keyboard('e');

    expect(func).toBeCalledTimes(1);

    userEvent.click(items.getByTestId('tobeclicked'));

    expect(func).toBeCalledTimes(2);
  });
});
