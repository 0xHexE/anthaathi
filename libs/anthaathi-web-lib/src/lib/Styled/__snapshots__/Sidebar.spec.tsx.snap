// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar should render the sidebar 1`] = `
<div
  aria-expanded="false"
  class=""
  data-testid="sidebar-title"
  role="button"
  tabindex="0"
>
  <div
    class=""
    data-baseweb="typo-labelsmall"
  >
    title
  </div>
  <span
    class=""
  />
  <span
    data-kind="action"
  >
    <button
      data-testid="hello"
      type="button"
    >
      Hello world
    </button>
  </span>
  <svg
    aria-hidden="true"
    fill="currentColor"
    focusable="false"
    height="16"
    preserveAspectRatio="xMidYMid meet"
    viewBox="0 0 16 16"
    width="16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M8 5L13 10 12.3 10.7 8 6.4 3.7 10.7 3 10z"
    />
  </svg>
</div>
`;
